common/src/main/kotlin/no/nav/su/se/bakover/common/CorrelationId.kt

```diff
@@ -1,6 +1,46 @@
 package no.nav.su.se.bakover.common
 
+import org.slf4j.MDC
+import java.util.UUID
+
+const val CorrelationIdHeader = "X-Correlation-ID"
+
 @JvmInline
 value class CorrelationId(val value: String) {
     override fun toString() = value
+
+    companion object {
+        fun generate(): CorrelationId = CorrelationId(UUID.randomUUID().toString())
+
+        /** Henter `X-Correlation-ID` fra MDC */
+        fun getCorrelationIdFromThreadLocal(): CorrelationId? {
+            return MDC.get(CorrelationIdHeader)?.let { CorrelationId(it) } ?: null.also {
+                log.error(
+                    "Mangler '$CorrelationIdHeader' på MDC. Er dette et asynk-kall? Da må det settes manuelt, så tidlig som mulig.",
+                    RuntimeException("Genererer en stacktrace for enklere debugging."),
+                )
+            }
+        }
+
+        /** Henter en CorrelationId eller genererer en ny tilfeldig hver gang denne funksjonen kalles. */
+        fun getOrCreateCorrelationIdFromThreadLocal(): String {
+            // TODO jah: Bytt til CorrelationId for de andre som bruker denne typen og.
+            return getCorrelationIdFromThreadLocal()?.toString() ?: generate().toString()
+        }
+
+//        fun createCorrelationIdOnThreadLocalIfMissing() {
+//            MDC.put(mdcName, generate().toString())
+//        }
+
+        /**
+         * Overwrites the current MDC value if it is set.
+         * Removes it after the block has finished running.
+         */
+        infix fun withCorrelationId(body: (CorrelationId) -> Unit) {
```

Comment:
Noe å tenke på her i forhold til 
Application.kt:238
`callIdMdc("X-Correlation-ID")`

Reply:
Også; hadde det vært bedre om man gjorde et forsøk på å finne en verdi før man overskrev? På denne måten slipper man å ta stilling til at man gjør nestede kall til denne funksjonen og veldig enkelt for konsumentene å forholde seg til (passer vel litt inn med forslaget mitt ang å abstrahere vekk om denne verdien eksisterer fra før for konsumentene)

Reply:
Denne er kun tenkt brukt fra asynk jobber/consumers. Ktor har kontroll på dette selv vha. `callIdMdc` hvor den også rydder opp. Det er ikke noe ønske om å gjenbruke `X-Correlation-ID` fra andre jobber, derfor må vi overskrive denne så tidlig som mulig når en jobb kjøres.

For coroutines kan man og bruke coroutineContext: https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/current-coroutine-context.html

Reply:
Legger på en litt bedre beskrivelse og wrapper det i en try-finally